go/vt/vtgate/balancer/session.go

```diff
@@ -0,0 +1,226 @@
+}
+
+// NewSessionBalancer creates a new session balancer.
+func NewSessionBalancer(ctx context.Context, localCell string, topoServer srvtopo.Server, hc discovery.HealthCheck) (TabletBalancer, error) {
+	log.Info("session balancer: creating new session balancer")
+
+	b := &SessionBalancer{
+		localCell:     localCell,
+		hc:            hc,
+		localRings:    make(map[discovery.KeyspaceShardTabletType]*hashRing),
+		externalRings: make(map[discovery.KeyspaceShardTabletType]*hashRing),
+	}
+
+	// Set up health check subscription
+	hcChan := b.hc.Subscribe("SessionBalancer")
+
+	// Build initial hash rings
+
+	// Find all the targets we're watching
+	targets, _, err := srvtopo.FindAllTargetsAndKeyspaces(ctx, topoServer, b.localCell, discovery.KeyspacesToWatch, slices.Collect(maps.Keys(tabletTypesToWatch)))
+	if err != nil {
+		return nil, fmt.Errorf("session balancer: failed to find all targets and keyspaces: %w", err)
+	}
+
+	// Add each tablet to the hash ring
+	for _, target := range targets {
+		tablets := b.hc.GetHealthyTabletStats(target)
+		for _, tablet := range tablets {
+			b.onTabletHealthChange(tablet)
+		}
+	}
+
+	// Start watcher to keep track of tablet health
+	go b.watchHealthCheck(ctx, hcChan)
+
+	return b, nil
+}
+
+// Pick is the main entry point to the balancer.
+//
+// For a given session, it will return the same tablet for its duration, with preference to tablets
+// in the local cell.
+func (b *SessionBalancer) Pick(target *querypb.Target, _ []*discovery.TabletHealth, opts PickOpts) *discovery.TabletHealth {
+	if opts.SessionUUID == "" {
+		return nil
+	}
+
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	// Try to find a tablet in the local cell first
+	tablet := getFromRing(b.localRings, target, opts.InvalidTablets, opts.SessionUUID)
+	if tablet != nil {
+		return tablet
+	}
+
+	// If we didn't find a tablet in the local cell, try external cells
+	tablet = getFromRing(b.externalRings, target, opts.InvalidTablets, opts.SessionUUID)
+	return tablet
+}
+
+// DebugHandler provides a summary of the session balancer state.
+func (b *SessionBalancer) DebugHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "text/plain")
+	fmt.Fprintf(w, "Session balancer\n")
+	fmt.Fprintf(w, "================\n")
+	fmt.Fprintf(w, "Local cell: %s\n\n", b.localCell)
+
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	fmt.Fprint(w, b.print())
+}
+
+// watchHealthCheck watches the health check channel for tablet health changes, and updates hash rings accordingly.
+func (b *SessionBalancer) watchHealthCheck(ctx context.Context, hcChan chan *discovery.TabletHealth) {
+	for {
+		select {
+		case <-ctx.Done():
+			b.hc.Unsubscribe(hcChan)
+			return
+		case tablet := <-hcChan:
+			if tablet == nil {
+				continue
+			}
+
+			// Ignore tablets we aren't supposed to watch
+			if _, ok := tabletTypesToWatch[tablet.Target.TabletType]; !ok {
+				continue
+			}
+
+			b.onTabletHealthChange(tablet)
+		}
+	}
+}
```

Comment:
I have concern with how we can managing active tablets list. We already do that in "healthcheck.go" and it is more robust of how to use the tablet health checks to track all the tablets for a tablet_type. We should piggy back on that logic to build the hash ring. Here already I can see the miss that if we added a replica to the hash ring and later that tablet gets promoted to primary it will not be removed from the hash ring and we will continue to send traffic to that.